File: interpreter-pattern/js/src/SubtractExpression.js
import { Expression } from './Expression.js'

// 具体表达式，实现了抽象表达式接口
export class SubtractExpression extends Expression {
  constructor(exprOne, exprTwo) {
    super()
    this.exprOne = exprOne
    this.exprTwo = exprTwo
  }

  // 覆盖表达式，执行context对象
  interpret(context) {
    console.log(
      this.constructor.name +
        '::interpret() [context = ' +
        context.constructor.name +
        ']',
    )
    return this.exprOne.interpret(context) - this.exprTwo.interpret(context)
  }
}

File: interpreter-pattern/c/src/subtract_expression.c
#include "func.h"

// 覆盖表达式，执行context对象
int subtract_expression_interpret(Expression *exp, Context *context)
{
  printf("\r\n SubExpression::interpret() [exp_one=%s, exp_two=%s]", exp->exp_one->key, exp->exp_two->key);
  int one_result = exp->exp_one->interpret(exp->exp_one, context);
  int two_result = exp->exp_two->interpret(exp->exp_two, context);
  return one_result - two_result;
}

SubtractExpression *subtract_expression_constructor(Expression *exp_one, Expression *exp_two)
{
  printf("\r\n sub_expression_constructor() [exp_one=%s, exp_two=%s]", exp_one->key, exp_two->key);
  Expression *exp = (Expression *)malloc(sizeof(Expression));
  exp->interpret = &subtract_expression_interpret;
  exp->exp_one = exp_one;
  exp->exp_two = exp_two;
  SubtractExpression *sub_expression = (SubtractExpression *)exp;
  return sub_expression;
}